Add tensor.to(options)

ssnl opened the pull request:
@ezyang on the template hack
@smessmer on SFINAE of the `TensorOptions(Device)`
@goldsborough on the C++ API test changes
@zdevito on the `jit` codegen changes

ssnl committed:
Add tensor.to(options)

ssnl committed:
update err msg

ssnl committed:
put away before fixing jit

ssnl committed:
Workaround ambiguity

ssnl committed:
GEN_TO_SOURCE

ssnl committed:
std::decay<T>

SsnL commented on aten/src/ATen/core/TensorOptions.h:
```diff
@@ -75,20 +121,30 @@ struct CAFFE2_API TensorOptions {
   /// Constructs a `TensorOptions` object with the given device.
-  /* implicit */ TensorOptions(Device device) : TensorOptions() {
+  /// See NOTE [ TensorOptions Constructors ] on why this is templatized.
+  template<typename T, typename = c10::guts::enable_if_t<std::is_same<T, Device>::value>>
```
Thanks! Added in new commit!

ssnl committed:
beautify the space

ssnl committed:
decay_t

ssnl committed:
add note ref

ssnl committed:
fix not keeping the current device when only dtype is specified

SsnL commented:
sigh this still breaks jit tracing

zdevito commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -254,8 +271,7 @@ def declkey(decl):
             if arg['simple_type'] == 'TensorOptions':
-                del arguments[n]
-                arguments.extend([
+                decl['arguments'] = arguments[:n] + [
```
This can be more easily understood as a flat map:
```
def  expand_options(i, a):
  if a['simple_type'] != 'TensorOptions':
    return [a]
  assert('tensor_options_arg_index' not in n)
  decl['tensor_options_arg_index'] = n
  return [ {...}, {...}, {...} ]
decl['arguments'] = [ a for x in expand_options(i, arg) for i, arg in enumerate(decl['arguments']) ]
```

SsnL replied:
Thanks for the suggestion! 

Is tracing disabled on methods with `TensorOptions` argument? `.to(device)` now dispatches to `at::Tensor::to(TensorOptions)` and it seems to be traced as a no-op. I haven't looked at the JIT generated code, but will do tomorrow.

zdevito replied:
Tracing should be enabled. We trace torch.zeros, for instance. Not sure why `to` doesn't trace then.

SsnL replied:
Alright then. I'll investigate tomorrow!

zdevito commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -91,15 +91,20 @@ def from_ivalue(arg, value):
+        .dtype(${dtype})
+        .layout(${layout})
+        .device(${device});
+DeviceGuard device_guard(deviceForInputs(stack, ${num_inputs}));
```
DeviceGuard is no longer needed here. I have another open PR that removes it from the other place in the file.

zdevito reviewed (commented):
Not a full review, just looked at the jit code.

ssnl committed:
Unify codegen detection on whether something is a factory method

Add more tests for copy behavior

Fix to(options) copies if device has unset index

Make ensure_has_index to return a copy now that Device is mall

smessmer commented:
The SFINAE stuff looks correct. I'm, however, not certain if we really want this API in C++. It seems a bit too magical to me. If we require people to define the type, it's much easier, for example:

    tensor.to(Device {kCUDA, 1});

SsnL commented:
@smessmer Yes, I agree that it is a bit too magical. The reason I did the SFINAE part is not for `tensor.to` but to support existing usage of constructing TensorOptions via `TensorOptions({kCUDA, -1})`. Maybe we don't necessarily need to support that, but the error message will be weird.. cc @goldsborough

ssnl committed:
address comments from @zdevito

ssnl committed:
add BC for to(TypeMeta)

ssnl committed:
Use typeMetaToScalarType

ssnl committed:
change core/Tensor.h

ezyang commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -284,25 +300,27 @@ def declkey(decl):
     jit_decls = [d for d in aten_decls if is_jit_op(d)]
 
     # add arguments dtype and device for functions like zeros
+    def expand_options(decl, i, arg):
```
Was there any substantive change here?

ezyang replied:
Oh I see, you no longer set `has_tensor_options`.

ezyang reviewed (approved):
Thanks a lot! It's kind of complicated but I don't have a better solution for you.
